File: imogi_finance/imogi_finance/report/withholding_register/withholding_register.py
"""
Withholding Register - ERPNext v15+ Native-First

Shows GL Entry based withholding tax transactions.
Only includes valid, non-cancelled GL entries for configured PPh accounts.
"""

from __future__ import annotations

import frappe
from frappe import _
from frappe.query_builder import DocType
from frappe.query_builder.functions import Sum, Coalesce
from frappe.utils import flt, getdate
from typing import Optional, Dict, List, Any

from imogi_finance.imogi_finance.utils.tax_report_utils import (
	validate_withholding_configuration,
	get_pph_accounts_for_company,
	build_date_conditions,
	get_columns_with_width
)


def execute(filters: Optional[Dict[str, Any]] = None) -> tuple[List[Dict], List[Dict]]:
	"""
	Execute Withholding Register report.
	
	Args:
		filters: Report filters
		
	Returns:
		Tuple of (columns, data)
	"""
	filters = filters or {}
	
	# Company is required for this report
	company = filters.get("company")
	if not company:
		frappe.throw(_("Company is required for Withholding Register"))
	
	# Validate configuration
	validation = validate_withholding_configuration(company)
	if not validation.get("valid"):
		frappe.msgprint(
			f"{validation.get('message')}<br>{validation.get('action', '')}",
			title=_("Configuration Error"),
			indicator=validation.get("indicator", "red"),
			raise_exception=True
		)
	
	# Get accounts to filter - either from filter or from Tax Profile
	accounts = filters.get("accounts")
	if accounts and not isinstance(accounts, list):
		accounts = [accounts]
	
	if not accounts:
		accounts = validation.get("accounts", [])
	
	if not accounts:
		frappe.throw(_("No withholding tax accounts configured or selected"))
	
	columns = get_columns()
	data = get_data(filters, company, accounts)
	
	return columns, data


def get_columns() -> List[Dict[str, Any]]:
	"""Define report columns with ERPNext v15+ standards."""
	columns = [
		{
			"label": _("Posting Date"),
			"fieldname": "posting_date",
			"fieldtype": "Date",
			"width": 100
		},
		{
			"label": _("Account"),
			"fieldname": "account",
			"fieldtype": "Link",
			"options": "Account",
			"width": 200
		},
		{
			"label": _("Party Type"),
			"fieldname": "party_type",
			"fieldtype": "Data",
			"width": 120
		},
		{
			"label": _("Party"),
			"fieldname": "party",
			"fieldtype": "Dynamic Link",
			"options": "party_type",
			"width": 180
		},
		{
			"label": _("Voucher Type"),
			"fieldname": "voucher_type",
			"fieldtype": "Data",
			"width": 140
		},
		{
			"label": _("Voucher No"),
			"fieldname": "voucher_no",
			"fieldtype": "Dynamic Link",
			"options": "voucher_type",
			"width": 160
		},
		{
			"label": _("Debit"),
			"fieldname": "debit",
			"fieldtype": "Currency",
			"width": 120
		},
		{
			"label": _("Credit"),
			"fieldname": "credit",
			"fieldtype": "Currency",
			"width": 120
		},
		{
			"label": _("Net Amount"),
			"fieldname": "net_amount",
			"fieldtype": "Currency",
			"width": 130
		},
		{
			"label": _("Remarks"),
			"fieldname": "remarks",
			"fieldtype": "Small Text",
			"width": 220
		},
	]
	
	return get_columns_with_width(columns)


def get_data(filters: Dict[str, Any], company: str, accounts: List[str]) -> List[Dict[str, Any]]:
	"""
	Get Withholding Register data using frappe.qb.
	Returns only valid, non-cancelled GL entries.
	"""
	GL = DocType("GL Entry")
	
	# Build base query - only non-cancelled GL entries
	query = (
		frappe.qb.from_(GL)
		.select(
			GL.posting_date,
			GL.account,
			GL.party_type,
			GL.party,
			GL.voucher_type,
			GL.voucher_no,
			GL.debit,
			GL.credit,
			GL.remarks
		)
		.where(GL.company == company)
		.where(GL.is_cancelled == 0)
	)
	
	# Filter by accounts
	if accounts:
		query = query.where(GL.account.isin(accounts))
	
	# Apply date range
	date_condition = build_date_conditions(GL, filters, "posting_date")
	if date_condition is not None:
		query = query.where(date_condition)
	
	# Filter by party if specified
	if filters.get("party"):
		query = query.where(GL.party == filters.get("party"))
	
	# Filter by voucher type if specified
	if filters.get("voucher_type"):
		query = query.where(GL.voucher_type == filters.get("voucher_type"))
	
	# Order by posting date and account
	query = query.orderby(GL.posting_date).orderby(GL.account).orderby(GL.voucher_no)
	
	# Execute query
	entries = query.run(as_dict=True)
	
	# Calculate net amount (credit - debit for liability accounts)
	for entry in entries:
		entry["net_amount"] = flt(entry.get("credit", 0)) - flt(entry.get("debit", 0))
	
	return entries

File: imogi_finance/imogi_finance/report/withholding_register/withholding_register.js
// Copyright (c) 2024, Imogi Finance and contributors
// For license information, please see license.txt

frappe.query_reports["Withholding Register"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_start(),
			"reqd": 0
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.month_end(),
			"reqd": 0
		},
		{
			"fieldname": "accounts",
			"label": __("Account"),
			"fieldtype": "Link",
			"options": "Account",
			"get_query": function() {
				const company = frappe.query_report.get_filter_value("company");
				return {
					filters: {
						"company": company,
						"account_type": ["in", ["Tax", "Payable"]],
						"is_group": 0
					}
				};
			}
		},
		{
			"fieldname": "party",
			"label": __("Party"),
			"fieldtype": "Link",
			"options": "Supplier",
			"get_query": function() {
				return {
					query: "erpnext.controllers.queries.supplier_query"
				};
			}
		},
		{
			"fieldname": "voucher_type",
			"label": __("Voucher Type"),
			"fieldtype": "Select",
			"options": ["", "Purchase Invoice", "Payment Entry", "Journal Entry", "Expense Claim"]
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Format net amount with color indicators
		if (column.fieldname == "net_amount") {
			const amount = parseFloat(data.net_amount);
			if (amount > 0) {
				// Credit balance (liability increased)
				value = `<span style="color: var(--text-green)">${value}</span>`;
			} else if (amount < 0) {
				// Debit balance (liability decreased/payment)
				value = `<span style="color: var(--text-red)">${value}</span>`;
			}
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Validate configuration on load
		const company = frappe.query_report.get_filter_value("company");
		if (company) {
			frappe.call({
				method: "imogi_finance.imogi_finance.utils.tax_report_utils.validate_tax_register_configuration",
				args: {
					register_type: "withholding",
					company: company
				},
				callback: function(r) {
					if (r.message && !r.message.valid) {
						frappe.msgprint({
							title: __("Configuration Required"),
							indicator: r.message.indicator || "red",
							message: r.message.message + "<br>" + (r.message.action || "")
						});
					}
				}
			});
		}
		
		// Add custom buttons
		report.page.add_inner_button(__("Configuration Check"), function() {
			check_configuration(report);
		}, __("Tools"));
		
		report.page.add_inner_button(__("Show PPh Accounts"), function() {
			show_pph_accounts(report);
		}, __("Tools"));
		
		report.page.add_inner_button(__("Export to Excel"), function() {
			export_to_excel(report);
		}, __("Tools"));
	}
};

function check_configuration(report) {
	const company = frappe.query_report.get_filter_value("company");
	
	if (!company) {
		frappe.msgprint(__("Please select a company first."));
		return;
	}
	
	frappe.call({
		method: "imogi_finance.imogi_finance.utils.tax_report_utils.validate_tax_register_configuration",
		args: {
			register_type: "withholding",
			company: company
		},
		callback: function(r) {
			if (r.message) {
				let message = r.message.message;
				if (r.message.action) {
					message += "<br><br>" + r.message.action;
				}
				
				// Show configured accounts if available
				if (r.message.accounts && r.message.accounts.length > 0) {
					message += "<br><br><strong>" + __("Configured PPh Accounts:") + "</strong><br>";
					message += "<ul>";
					r.message.accounts.forEach(function(account) {
						message += `<li>${account}</li>`;
					});
					message += "</ul>";
				}
				
				frappe.msgprint({
					title: __("Configuration Status"),
					indicator: r.message.indicator || "blue",
					message: message
				});
			}
		}
	});
}

function show_pph_accounts(report) {
	const company = frappe.query_report.get_filter_value("company");
	
	if (!company) {
		frappe.msgprint(__("Please select a company first."));
		return;
	}
	
	// Get Tax Profile for the company
	frappe.db.get_value("Tax Profile", {"company": company}, "name", function(r) {
		if (r && r.name) {
			frappe.set_route("Form", "Tax Profile", r.name);
		} else {
			frappe.msgprint({
				title: __("Tax Profile Not Found"),
				indicator: "red",
				message: __("No Tax Profile found for company {0}. Please create one to configure PPh accounts.", [company])
			});
		}
	});
}

function export_to_excel(report) {
	// Use native ERPNext export functionality
	const filters = frappe.query_report.get_filter_values();
	
	frappe.call({
		method: "frappe.desk.query_report.export_query",
		args: {
			report_name: report.report_name,
			file_format_type: "Excel",
			filters: filters
		},
		callback: function(r) {
			if (r.message) {
				const link = document.createElement("a");
				link.href = r.message.file_url;
				link.download = r.message.file_name;
				link.click();
			}
		}
	});
}
